### PR TITLE
Reworked ViewComment/ViewTopic

### PR DESCRIPTION
**Pull request by fullstackly**
... made constructors for both and removed CommendsIDs field from DB generating it on synthetic struct ViewTopic

### CODE REVIEW

File: pkg/controllers/forum.go [hunk 1]
```diff
@@ -29,7 +29,11 @@ func (c *Controller) TopicsGetHandler() http.HandlerFunc {
 			if err != nil {
 				logger.Error(err)
 			}
-			viewTopics = append(viewTopics, forum.ViewTopic{userName, topic})
+			viewTopic, err := c.ForumStore.NewViewTopic(userName, topic)
+			if err != nil {
```
Comment (yudinRazDvaTri) [inline pkg/controllers/forum.go]: What is the point of this error handling?

Comment (GorgeousPuree) [inline pkg/controllers/forum.go:33]: What is the point of this error handling?

File: pkg/controllers/forum.go [hunk 2]
```diff
@@ -83,11 +87,11 @@ func (c *Controller) CommentsGetHandler() http.HandlerFunc {
 			if err != nil {
```
Comment (GorgeousPuree) [inline pkg/controllers/forum.go:87]: What is the point of this error handling?

File: pkg/controllers/forum.go [hunk 3]
```diff
@@ -83,11 +87,11 @@ func (c *Controller) CommentsGetHandler() http.HandlerFunc {
...
 				logger.Error(err)
 			}
-			ratedUsers, err := c.ForumStore.GetCommentRatings(comment.CommentID)
+			viewComment, err := c.ForumStore.NewViewComment(userName, comment)
 			if err != nil {
```
Comment (GorgeousPuree) [inline pkg/controllers/forum.go:91]: What is the point of this error handling?

File: pkg/controllers/search.go [hunk 4]
```diff
@@ -97,7 +97,11 @@ func(c *Controller) searchByForum(w http.ResponseWriter, r * http.Request){
 			if err != nil {
 				logger.Error(err)
 			}
-			viewTopics = append(viewTopics, forum.ViewTopic{userName, topic})
+			viewTopic, err := c.ForumStore.NewViewTopic(userName, topic)
+			if err != nil {
```
Comment (GorgeousPuree) [inline pkg/controllers/search.go:101]: What is the point of this error handling?

File: pkg/controllers/search.go [hunk 5]
```diff
@@ -114,7 +118,11 @@ func(c *Controller) searchByForum(w http.ResponseWriter, r * http.Request){
 			if err != nil {
```
Comment (GorgeousPuree) [inline pkg/controllers/search.go:118]: What is the point of this error handling?

File: pkg/controllers/search.go [hunk 6]
```diff
@@ -114,7 +118,11 @@ func(c *Controller) searchByForum(w http.ResponseWriter, r * http.Request){
...
 				logger.Error(err)
 			}
-			viewTopics = append(viewTopics, forum.ViewTopic{userName, topic})
+			viewTopic, err := c.ForumStore.NewViewTopic(userName, topic)
+			if err != nil {
```
Comment (GorgeousPuree) [inline pkg/controllers/search.go:122]: What is the point of this error handling?

File: pkg/controllers/forum.go [hunk 7]
```diff
@@ -29,7 +29,11 @@ func (c *Controller) TopicsGetHandler() http.HandlerFunc {
...
```
Comment (GorgeousPuree) [inline pkg/controllers/forum.go:29]: What is the point of this error handling?

File: pkg/controllers/search.go [hunk 8]
```diff
@@ -97,7 +97,11 @@ func(c *Controller) searchByForum(w http.ResponseWriter, r * http.Request){
 			if err != nil {
```
Comment (GorgeousPuree) [inline pkg/controllers/search.go:97]: What is the point of this error handling?

File: pkg/models/forum/comment.go [hunk 9]
```diff
@@ -75,21 +88,39 @@ func (f *ForumStore) RateComment(commentID, userID int) (bool, error) {
...
 		err = fmt.Errorf("Can't read rating-rows from db: %v", err)
 	}
 	defer rows.Close()
 
 	for rows.Next() {
-		var like int64
-		err = rows.Scan(&like)
+		var likerID int64
+		err = rows.Scan(&likerID)
 		if err != nil {
```
Comment (GorgeousPuree) [inline pkg/models/forum/comment.go:101]: Why is there no return statement? Why do application need to continue work in this function, if error happens?

**Comment by fullstackly** [comment]
This is only a cropped piece of code. Look-up for original forum.go's
TopicsGetHandler to see full code.

File: pkg/models/forum/comment.go [hunk 10]
```diff
@@ -75,21 +88,39 @@ func (f *ForumStore) RateComment(commentID, userID int) (bool, error) {
 	return rateOk, err
 }
 
-func (f *ForumStore) GetCommentRatings(commentID int) (likes []int64, err error) {
+func (f *ForumStore) getCommentLikers(commentID int) (likersIDs []int64, err error) {
 	rows, err := f.DB.Query(`SELECT user_id FROM ratings WHERE comment_id = $1;`, commentID)
 	if err != nil {
```
Comment (GorgeousPuree) [inline pkg/models/forum/comment.go:93]: Where is return statement? If sql query fails, application will throw error caused by referencing to invalid memory address or nil pointer dereference.

File: pkg/models/forum/comment.go [hunk 11]
```diff
@@ -75,21 +88,39 @@ func (f *ForumStore) RateComment(commentID, userID int) (bool, error) {
...
+		likersIDs = append(likersIDs, likerID)
 	}
-
 	return
 }
+
+func (f *ForumStore) getCommentsIDs(topicID int) (commentsIDs []int64, err error) {
+	rows, err := f.DB.Query(
+		`SELECT comment_id FROM comments WHERE topic_id = $1;`, topicID)
+	if err != nil {
```
Comment (GorgeousPuree) [inline pkg/models/forum/comment.go:112]: Where is return statement? If sql query fails, application will throw error caused by referencing to invalid memory address or nil pointer dereference.

File: pkg/models/forum/comment.go [hunk 12]
```diff
@@ -75,21 +88,39 @@ func (f *ForumStore) RateComment(commentID, userID int) (bool, error) {
...
+		err = fmt.Errorf("Can't number of comments in topic %d from DB: %v", topicID, err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var commentID int64
+		err = rows.Scan(&commentID)
+		if err != nil {
```
Comment (GorgeousPuree) [inline pkg/models/forum/comment.go:120]: Why is there no return statement? Why do application need to continue work in this function, if error happens?

File: pkg/models/forum/topic.go [hunk 13]
```diff
@@ -14,18 +13,28 @@ type Topic struct {
...
+		viewTopic = ViewTopic{userName, commentsIDs, topic}
+		return
+}
+
 func (f *ForumStore) CreateNewTopic(userID int, title, description string) (err error) {
-	_, err = f.DB.Exec(
-		`insert into topics (user_id, title, description) values ($1, $2, $3)`,
-		userID, title, description)
+	_, err = f.DB.Exec(`insert into topics (user_id, title, description) values ($1, $2, $3)`, userID, title, description)
 	if err != nil {
```
Comment (GorgeousPuree) [inline pkg/models/forum/topic.go:38]: Why is there no return statement?

File: pkg/models/forum/topic.go [hunk 14]
```diff
@@ -43,12 +52,6 @@ func (f *ForumStore) GetAllTopics() (topics []Topic, err error) {
...
```
Comment (GorgeousPuree) [inline pkg/models/forum/topic.go:52]: Why is there no return statement?

File: pkg/models/forum/topic.go [hunk 15]
```diff
@@ -66,25 +69,8 @@ func (f *ForumStore) GetTopicComments(topicID int) (comments []Comment, err erro
...
-
 func (f *ForumStore) GetTopicByID(topicID int) (topic Topic, err error) {
-	err = f.SetNumberOfComments(topicID)
-	if err != nil {
-		logger.Error(err)
-	}
-	err = f.DB.QueryRowx(
-		`SELECT * FROM topics WHERE topic_id = $1`, topicID).StructScan(&topic)
+	err = f.DB.QueryRowx(`SELECT * FROM topics WHERE topic_id = $1`, topicID).StructScan(&topic)
 	if err != nil {
```
Comment (GorgeousPuree) [inline pkg/models/forum/topic.go:74]: Why is there no return statement in if statement?

**Review by GorgeousPuree — CHANGES_REQUESTED** [review]
Check out the comments. Add "return" in your if-statements, if error is not null. If you would like to make user-friendly notifications (like "post is not found"), then make custom errors in package utilerr and use them. Please do not push your changes if you know that they are need to be refactored.

**Comment by GorgeousPuree** [comment]
> This is only a cropped piece of code. Look-up for original forum.go's TopicsGetHandler to see full code.

Those error handlers in controllers are doing nothing.

**Pull request closed without merging**